File: examples/data_structures/stack.py
# @depyler: thread_safety = "required"
# @depyler: bounds_checking = "explicit"
from typing import List, Optional

class Stack:
    """Thread-safe stack implementation"""
    
    def __init__(self) -> None:
        self._items: List[int] = []
    
    def push(self, item: int) -> None:
        """Push item onto stack"""
        self._items.append(item)
    
    def pop(self) -> Optional[int]:
        """Pop item from stack, return None if empty"""
        if self.is_empty():
            return None
        return self._items.pop()
    
    def peek(self) -> Optional[int]:
        """Look at top item without removing it"""
        if self.is_empty():
            return None
        return self._items[-1]
    
    def is_empty(self) -> bool:
        """Check if stack is empty"""
        return len(self._items) == 0
    
    def size(self) -> int:
        """Get number of items in stack"""
        return len(self._items)

def balanced_parentheses(expression: str) -> bool:
    """Check if parentheses are balanced using a stack"""
    stack = Stack()
    opening = "({["
    closing = ")}]"
    pairs = {"(": ")", "{": "}", "[": "]"}
    
    for char in expression:
        if char in opening:
            stack.push(ord(char))  # Store ASCII value
        elif char in closing:
            if stack.is_empty():
                return False
            
            last = stack.pop()
            if last is None:
                return False
            
            expected = ord(pairs[chr(last)])
            if ord(char) != expected:
                return False
    
    return stack.is_empty()

File: examples/data_structures/stack.rs
use std::collections::HashMap;
#[derive(Debug, Clone)]
pub struct IndexError {
    message: String,
}
impl std::fmt::Display for IndexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "index out of range: {}", self.message)
    }
}
impl std::error::Error for IndexError {}
impl IndexError {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}
#[derive(Debug, Clone)]
pub struct Stack {
    pub _items: Vec<i32>,
}
impl Stack {
    pub fn new() -> Self {
        Self { _items: Vec::new() }
    }
    pub fn push(&self, item: i32) {
        self._items.push(item);
    }
    pub fn pop(&self) -> Option<i32> {
        if self.is_empty() {
            return ();
        };
        return self._items.pop().unwrap_or_default();
    }
    pub fn peek(&self) -> Option<i32> {
        if self.is_empty() {
            return ();
        };
        return self._items[-1 as usize];
    }
    pub fn is_empty(&self) -> bool {
        return self._items.len() as i32 == 0;
    }
    pub fn size(&self) -> i32 {
        return self._items.len() as i32;
    }
}
#[doc = "Check if parentheses are balanced using a stack"]
pub fn balanced_parentheses(expression: &str) -> Result<bool, Box<dyn std::error::Error>> {
    let mut stack = Stack::new();
    let opening = "({[";
    let closing = ")}]";
    let pairs = {
        let mut map = HashMap::new();
        map.insert("(".to_string(), ")");
        map.insert("{".to_string(), "}");
        map.insert("[".to_string(), "]");
        map
    };
    for char in expression.chars() {
        if opening.contains(&*char) {
            stack.push(char.chars().next().unwrap() as i32);
        } else {
            if closing.contains(&*char) {
                if stack.is_empty() {
                    return Ok(false);
                }
                let last = stack.pop();
                if last.is_none() {
                    return Ok(false);
                }
                let expected = pairs
                    .get(&char::from_u32(last as u32).unwrap().to_string())
                    .cloned()
                    .unwrap_or_default()
                    .chars()
                    .next()
                    .unwrap() as i32;
                if char.chars().next().unwrap() as i32 != expected {
                    return Ok(false);
                }
            }
        }
    }
    Ok(stack.is_empty())
}
#[cfg(test)]
mod tests {
    use super::*;
    use quickcheck::{quickcheck, TestResult};
    #[test]
    fn test_balanced_parentheses_examples() {
        let _ = balanced_parentheses(Default::default());
    }
}
